foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/cache/VersionedCache.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2017 Huawei Technologies Co., Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.foundation.common.cache;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@SuppressWarnings("unchecked")
+public class VersionedCache {
+  protected static final AtomicInteger VERSION = new AtomicInteger();
+
+  interface IsEmpty {
+    boolean isEmpty();
+  }
+
+  protected int cacheVersion;
+
+  // an optional name
+  protected String name;
+
+  protected Object data;
+
+  protected IsEmpty isEmpty = this::isCommonEmpty;
+
+  private boolean isCommonEmpty() {
+    return data == null;
+  }
+
+  private boolean isMapEmpty() {
```

Comment:
If the data is not a map , we will get the class case exception. 
I don't like these three method.

Reply:
yes, i don't like this too.
but data in cache is not always the same type.
eg: in discoveryTree, before endpointFilter, discoveryTreeNode data is Map<String, MicroserviceInstance>, after endpointFilter, discoverTreeNode data is Collection<Endpoint> or Collection<CseServer>

caller know what's data type in cache, and call different method, if invoke wrong method, then it's a bug, developers should fix it.

Reply:
isXXXEmpty methods have no problems, because they are private, and used by data type.
not so good methods are: mapData/collectionData/arrayData